packages/components/src/image_zoom/hooks.jsx

```diff
@@ -0,0 +1,123 @@
+import { useEffect, useLayoutEffect, useMemo, useRef, useState, useCallback } from 'react';
+import {
+  getElementWidth,
+  getInsideBoundaries,
+  getOffsetForNewScale,
+  getOffsetForNewScaleWithCustomAnchor,
+} from './utils';
+
+export const usePrevious = (value) => {
+  const ref = useRef(null);
+  useEffect(() => { ref.current = value; });
+
+  return ref.current;
+};
+
+export const useImage = (src) => {
+  const [image, setImage] = useState(null);
+
+  useEffect(() => {
+    const imageObj = new Image(100, 100);
+    imageObj.src = src;
+    imageObj.onload = () => { setImage(imageObj); };
+  }, [src]);
+
+  return image;
+};
+
+export const useZoomHandler = (setScale, setOffset, previewSize, prevScale) => (
+  useCallback((scale, midpoint) => {
+    setScale(scale);
+    setOffset(({ top, left }) => {
+      console.group('newTop');
+      let newTop = getOffsetForNewScaleWithCustomAnchor(midpoint.y, top, prevScale, scale, previewSize.height);
+      console.groupEnd();
+      console.group('newLeft');
+      let newLeft = getOffsetForNewScaleWithCustomAnchor(midpoint.x, left, prevScale, scale, previewSize.width);
+      console.groupEnd();
+      return ({
+        top: newTop,
+        left: newLeft,
+      });
+    });
+  }, [prevScale, previewSize])
+);
+
+export const useOffsetUpdate = (setOffset, previewSize, scale) => {
+  const prevScale = usePrevious(scale);
+
+  useEffect(() => {
+    if (scale === null || prevScale === null) return;
+
+    setOffset(({ top, left }) => ({
+      top: getOffsetForNewScale(top, prevScale, scale, previewSize.height),
+      left: getOffsetForNewScale(left, prevScale, scale, previewSize.width),
+    }));
+  }, [scale]);
+};
+
+export const useContainerWidth = (ref) => {
+  const [width, setWidth] = useState(getElementWidth(ref?.current));
+
+  const handleResize = () => {
+    if (!ref.current) return;
+
+    setWidth(getElementWidth(ref.current));
+  };
+
+  useLayoutEffect(() => {
+    const el = ref.current;
+    if (!el) return;
+
+    handleResize();
+
+    const resizeObserver = new ResizeObserver(() => handleResize);
```

Comment:
resize observer doesn't work in some mobile safari versions - add polyfill in apps?